# Importing libraries

In [39]:
#import libraries
import pandas as pd
import numpy as np
import os

In [40]:
#import dataframe
ords_prods_merge = pd.read_pickle(r'/Users/lordsonobire/Documents/Data Immersion/Achievement 4/Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_merge_derived.pkl')

In [41]:
ords_prods_merge.shape

(30328763, 18)

In [42]:
#check to see if its right data frame and contains newly derived columns
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2398795        1             2                  3                  7   
1   2398795        1             2                  3                  7   
2   2398795        1             2                  3                  7   
3   2398795        1             2                  3                  7   
4   2398795        1             2                  3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    15.0       10258                  2          0   
2                    15.0       12427                  3          1   
3                    15.0       13176                  4          0   
4                    15.0       26088                  5          1   

                 product_name  aisle_id  department_id  prices match  \
0                        Soda        77              7     9.0  both   
1                  Pistachios       117             19     3.0  both   
2         Original Beef Jerky        23             19     4.4  both   
3      Bag of Organic Bananas        24              4    10.3  both   
4  Aged White Cheddar Popcorn        23             19     4.7  both   

     price_range_loc     busiest_day busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy   Least busy        Average orders  
1  Low-range product  Regularly busy   Least busy        Average orders  
2  Low-range product  Regularly busy   Least busy        Average orders  
3  Mid-range product  Regularly busy   Least busy        Average orders  
4  Low-range product  Regularly busy   Least busy        Average orders

In [43]:
#find the aggregated mean of the “order_number” column grouped by “department_id”
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                16.559358
2                18.413176
3                18.279600
4                18.915890
5                16.497751
6                17.609390
7                18.303975
8                16.383301
9                17.022963
10               21.227447
11               17.311768
12               16.953613
13               17.727470
14               17.898097
15               17.290123
16               18.757796
17               16.808683
18               20.336389
19               18.258620
20               17.578959
21               23.644477

# How do these results from the subset compare against results from the entire dataframe 
The mean 'order_number' values for the subset are generally lower compared to that of the entire dataset.
Department 21 has the highest mean value in both data sets.
The entire dataset provides a more accurate representation of the 'order_number' accross all departments by representing the overall order pattern than the subset which is likely narrower in scope and excludes certain transactions.

In [ ]:
ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [45]:
# Aggregating ords_prods_merge with the transform function 
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/g0/3g_bmqwx1ks0wzpxgh5vgbfm0000gn/T/ipykernel_60805/4180060956.py:2: FutureWarning: The provided callable <function max at 0x110c3a660> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [46]:
#checking output
ords_prods_merge.head(100)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2398795        1             2                  3                  7   
1    2398795        1             2                  3                  7   
2    2398795        1             2                  3                  7   
3    2398795        1             2                  3                  7   
4    2398795        1             2                  3                  7   
..       ...      ...           ...                ...                ...   
95   1199898        2             6                  2                  9   
96   1199898        2             6                  2                  9   
97   1199898        2             6                  2                  9   
98   1199898        2             6                  2                  9   
99   1199898        2             6                  2                  9   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     15.0         196                  1          1   
1                     15.0       10258                  2          0   
2                     15.0       12427                  3          1   
3                     15.0       13176                  4          0   
4                     15.0       26088                  5          1   
..                     ...         ...                ...        ...   
95                    13.0       32052                  5          0   
96                    13.0       27344                  6          1   
97                    13.0       24852                  7          1   
98                    13.0       45066                  8          1   
99                    13.0       27966                  9          1   

                         product_name  aisle_id  department_id  prices match  \
0                                Soda        77              7     9.0  both   
1                          Pistachios       117             19     3.0  both   
2                 Original Beef Jerky        23             19     4.4  both   
3              Bag of Organic Bananas        24              4    10.3  both   
4          Aged White Cheddar Popcorn        23             19     4.7  both   
..                                ...       ...            ...     ...   ...   
95  Organic Brown Rice Cake Salt-Free        78             19     4.3  both   
96               Uncured Genoa Salami        96             20    14.8  both   
97                             Banana        24              4    12.3  both   
98                   Honeycrisp Apple        24              4     2.5  both   
99                Organic Raspberries       123              4     4.4  both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy      Least busy        Average orders   
1   Low-range product  Regularly busy      Least busy        Average orders   
2   Low-range product  Regularly busy      Least busy        Average orders   
3   Mid-range product  Regularly busy      Least busy        Average orders   
4   Low-range product  Regularly busy      Least busy        Average orders   
..                ...             ...             ...                   ...   
95  Low-range product  Regularly busy  Regularly busy           Most orders   
96  Mid-range product  Regularly busy  Regularly busy           Most orders   
97  Mid-range product  Regularly busy  Regularly busy           Most orders   
98  Low-range product  Regularly busy  Regularly busy           Most orders   
99  Low-range product  Regularly busy  Regularly busy           Most orders   

    max_order  
0          10  
1          10  
2          10  
3          10  
4          10  
..        ...  
95         14  
96         14  
97         14  
98         14  
99         14  

[100 rows x 19 columns]

In [47]:
#create a loyalty label using loc function 
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [48]:
#examining the value counts on loyalty_flag
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15081691
Loyal customer      10095381
New customer         5151691
Name: count, dtype: int64

In [49]:
#checking updated df 
ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2398795        1             2                  3                  7   
1   2398795        1             2                  3                  7   
2   2398795        1             2                  3                  7   
3   2398795        1             2                  3                  7   
4   2398795        1             2                  3                  7   
5   2398795        1             2                  3                  7   
6    473747        1             3                  3                 12   
7    473747        1             3                  3                 12   
8    473747        1             3                  3                 12   
9    473747        1             3                  3                 12   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    15.0         196                  1          1   
1                    15.0       10258                  2          0   
2                    15.0       12427                  3          1   
3                    15.0       13176                  4          0   
4                    15.0       26088                  5          1   
5                    15.0       13032                  6          0   
6                    21.0         196                  1          1   
7                    21.0       12427                  2          1   
8                    21.0       10258                  3          1   
9                    21.0       25133                  4          0   

                 product_name  aisle_id  department_id  prices match  \
0                        Soda        77              7     9.0  both   
1                  Pistachios       117             19     3.0  both   
2         Original Beef Jerky        23             19     4.4  both   
3      Bag of Organic Bananas        24              4    10.3  both   
4  Aged White Cheddar Popcorn        23             19     4.7  both   
5       Cinnamon Toast Crunch       121             14     4.0  both   
6                        Soda        77              7     9.0  both   
7         Original Beef Jerky        23             19     4.4  both   
8                  Pistachios       117             19     3.0  both   
9       Organic String Cheese        21             16     8.6  both   

     price_range_loc     busiest_day busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy   Least busy        Average orders   
1  Low-range product  Regularly busy   Least busy        Average orders   
2  Low-range product  Regularly busy   Least busy        Average orders   
3  Mid-range product  Regularly busy   Least busy        Average orders   
4  Low-range product  Regularly busy   Least busy        Average orders   
5  Low-range product  Regularly busy   Least busy        Average orders   
6  Mid-range product  Regularly busy   Least busy           Most orders   
7  Low-range product  Regularly busy   Least busy           Most orders   
8  Low-range product  Regularly busy   Least busy           Most orders   
9  Mid-range product  Regularly busy   Least busy           Most orders   

   max_order  loyalty_flag  
0         10  New customer  
1         10  New customer  
2         10  New customer  
3         10  New customer  
4         10  New customer  
5         10  New customer  
6         10  New customer  
7         10  New customer  
8         10  New customer  
9         10  New customer

In [50]:
#Determining basic statistics of the products prices for each loyalty category
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.402162  1.0  99999.0
New customer      13.396333  1.0  99999.0
Regular customer  12.546842  1.0  99999.0

# Spending habits 
Loyal customers spend the least on average (10.40), new customers the most (13.40), and regular customers fall in between (12.55). This suggests loyal customers prefer affordability, while new customers explore higher-priced items. Further analysis could refine insights by addressing outliers like the maximum price of 99999.

In [64]:
#create a agg of the average prices accross orders 
ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/g0/3g_bmqwx1ks0wzpxgh5vgbfm0000gn/T/ipykernel_60805/3044266404.py:2: FutureWarning: The provided callable <function mean at 0x110c3b060> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [66]:
#checking the new column 
ords_prods_merge[['user_id', 'avg_product_price']].head(15)

user_id  avg_product_price
0         1           6.372222
1         1           6.372222
2         1           6.372222
3         1           6.372222
4         1           6.372222
5         1           6.372222
6         1           6.372222
7         1           6.372222
8         1           6.372222
9         1           6.372222
10        1           6.372222
11        1           6.372222
12        1           6.372222
13        1           6.372222
14        1           6.372222

In [70]:
# Creating a 'Spending' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [72]:
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [74]:
#examing the spending flag column 
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     29730505
High spender      598258
Name: count, dtype: int64

In [76]:
ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0         1           6.372222   Low spender
1         1           6.372222   Low spender
2         1           6.372222   Low spender
3         1           6.372222   Low spender
4         1           6.372222   Low spender
5         1           6.372222   Low spender
6         1           6.372222   Low spender
7         1           6.372222   Low spender
8         1           6.372222   Low spender
9         1           6.372222   Low spender
10        1           6.372222   Low spender
11        1           6.372222   Low spender
12        1           6.372222   Low spender
13        1           6.372222   Low spender
14        1           6.372222   Low spender

In [78]:
# Aggregating Data with transform() Function
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/g0/3g_bmqwx1ks0wzpxgh5vgbfm0000gn/T/ipykernel_60805/981590324.py:2: FutureWarning: The provided callable <function median at 0x111259d00> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [80]:
# Checking Relevant Columns 
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0         1                    15.0                        20.5
1         1                    15.0                        20.5
2         1                    15.0                        20.5
3         1                    15.0                        20.5
4         1                    15.0                        20.5
5         1                    15.0                        20.5
6         1                    21.0                        20.5
7         1                    21.0                        20.5
8         1                    21.0                        20.5
9         1                    21.0                        20.5
10        1                    21.0                        20.5
11        1                    29.0                        20.5
12        1                    29.0                        20.5
13        1                    29.0                        20.5
14        1                    29.0                        20.5

In [84]:
# Creating a 'Order Frequency' Label Flag using loc Function
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [86]:
# Examining Value Counts on 'order_frequency_flag'
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        20675290
Regular customer          6594542
Non-frequent customer     3058931
Name: count, dtype: int64

In [88]:
# Checking Relevant Columns of Updated Dataframe
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                    15.0                        20.5   
1         1                    15.0                        20.5   
2         1                    15.0                        20.5   
3         1                    15.0                        20.5   
4         1                    15.0                        20.5   
5         1                    15.0                        20.5   
6         1                    21.0                        20.5   
7         1                    21.0                        20.5   
8         1                    21.0                        20.5   
9         1                    21.0                        20.5   
10        1                    21.0                        20.5   
11        1                    29.0                        20.5   
12        1                    29.0                        20.5   
13        1                    29.0                        20.5   
14        1                    29.0                        20.5   

     order_frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9   Non-frequent customer  
10  Non-frequent customer  
11  Non-frequent customer  
12  Non-frequent customer  
13  Non-frequent customer  
14  Non-frequent customer

# Export 

In [93]:
ords_prods_merge.to_pickle('/Users/lordsonobire/Documents/Data Immersion/Achievement 4/Instacart Basket Analysis/02 Data/Prepared Data/ord_prods_grouped.pkl')